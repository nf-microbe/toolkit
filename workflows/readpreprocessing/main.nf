/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    IMPORT MODULES / SUBWORKFLOWS / FUNCTIONS
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

include { FASTQ_READPREPROCESSING_FASTQ } from '../../subworkflows/nf-core/fastq_readpreprocessing_fastq'
include { getGenomeAttribute            } from '../../subworkflows/local/utils_nfmicrobe_toolkit_pipeline'

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    RUN MAIN WORKFLOW
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

workflow READPREPROCESSING {

    take:
    input_fastqs    // channel: fastq files read in from --input (samplesheet) and/or --fastqs
    parameters      // pipeline parameters

    main:

    ch_versions             = Channel.empty()
    ch_multiqc_files        = Channel.empty()
    ch_workdirs_to_clean    = Channel.empty()

    // Prepare host fasta and bowtie2 index for host removal
    if (parameters.run_bowtie2_host_removal) {
        // Load igenomes fasta and bowtie2 index
        igenomes_fasta   = getGenomeAttribute('fasta')
        igenomes_index   = getGenomeAttribute('bowtie2')

        // create bowtie2 fasta and index channels based on input parameters
        if (parameters.genome) {
            ch_bowtie2_fasta    = file(igenomes_fasta, checkIfExists: true)
            ch_bowtie2_index    = file(igenomes_index, checkIfExists: true)
        } else if (parameters.host_bowtie2_index) {
            if (!parameters.host_fasta) {
                error "[nf-microbe/toolkit]: --run_bowtie2_host_removal = true but no iGenomes (--genome) or custom fasta (--host_fasta) provided"
            }
            ch_bowtie2_fasta    = file(parameters.host_fasta, checkIfExists: true)
            ch_bowtie2_index    = file(parameters.host_bowtie2_index, checkIfExists: true)
        } else {
            if (!parameters.host_fasta) {
                error "[nf-microbe/toolkit]: --run_bowtie2_host_removal = true but no iGenomes (--genome) or custom fasta (--host_fasta) provided"
            }
            ch_bowtie2_fasta    = file(parameters.host_fasta, checkIfExists: true)
            ch_bowtie2_index    = null
        }
    } else {
        ch_bowtie2_fasta    = null
        ch_bowtie2_index    = null
    }

    //
    // SUBWORKFLOW: Read preprocessing
    //
    FASTQ_READPREPROCESSING_FASTQ(
        input_fastqs,                           // channel: [ [ meta.id, meta.single_end, meta.run ], [ reads_1.fastq.gz, reads_2.fastq.gz ] ] (MANDATORY)
        parameters.run_fastp,                   // boolean: run fastp
        parameters.perform_run_merging,         // boolean: merge runs within samples
        parameters.run_bowtie2_host_removal,    // boolean: run host removal
        ch_bowtie2_fasta,                       // string: /path/to/fasta.fasta.gz (OPTIONAL)
        ch_bowtie2_index                        // string: /path/to/host_index.bt2 (OPTIONAL)
    )

    emit:
    preprocessed_fastq_gz   = FASTQ_READPREPROCESSING_FASTQ.out.preprocessed_fastq_gz
    multiqc_files           = ch_multiqc_files.mix(FASTQ_READPREPROCESSING_FASTQ.out.multiqc_files)
    versions                = FASTQ_READPREPROCESSING_FASTQ.out.versions
    workdirs_to_clean       = FASTQ_READPREPROCESSING_FASTQ.out.workdirs_to_clean
}

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    THE END
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/
