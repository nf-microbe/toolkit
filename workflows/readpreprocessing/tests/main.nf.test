nextflow_workflow {

    name "Test Subworkflow: READPREPROCESSING"
    script "../main.nf"
    workflow "READPREPROCESSING"

    tag "nfmicrobe_toolkit_ci"

    test("host_removal: igenomes input") {
        when {
            params {
                run_bowtie2_host_removal    = true
                genome                      = "EB1"
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test1', run:'1', single_end:false ], // meta map
                        [
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkifExists: true),
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkifExists: true)
                        ]
                    ],
                    [
                        [ id:'test2', run:'1', single_end:false ], // meta map
                        [
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test2_1.fastq.gz', checkifExists: true),
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test2_2.fastq.gz', checkifExists: true)
                        ]
                    ]
                )
                input[1] = params
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.versions
                    ).match()
                }
            )
        }
    }

    test("host_removal: local index input") {
        when {
            params {
                run_bowtie2_host_removal    = true
                host_bowtie2_index          = "${projectDir}/assets/test-datasets/bowtie2/"
                host_fasta                  = "${projectDir}/assets/test-datasets/fastas/genome.fasta.gz"
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test1', run:'1', single_end:false ], // meta map
                        [
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkifExists: true),
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkifExists: true)
                        ]
                    ],
                    [
                        [ id:'test2', run:'1', single_end:false ], // meta map
                        [
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test2_1.fastq.gz', checkifExists: true),
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test2_2.fastq.gz', checkifExists: true)
                        ]
                    ]
                )
                input[1] = params
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.versions
                    ).match()
                }
            )
        }
    }

    test("host_removal: local fasta input") {
        when {
            params {
                run_bowtie2_host_removal    = true
                host_fasta                  = "${projectDir}/assets/test-datasets/fastas/genome.fasta.gz"
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test1', run:'1', single_end:false ], // meta map
                        [
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkifExists: true),
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkifExists: true)
                        ]
                    ],
                    [
                        [ id:'test2', run:'1', single_end:false ], // meta map
                        [
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test2_1.fastq.gz', checkifExists: true),
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test2_2.fastq.gz', checkifExists: true)
                        ]
                    ]
                )
                input[1] = params
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.versions
                    ).match()
                }
            )
        }
    }

    test("host_removal: no host genome") {
        when {
            params {
                run_bowtie2_host_removal    = true
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test1', run:'1', single_end:false ], // meta map
                        [
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkifExists: true),
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkifExists: true)
                        ]
                    ],
                    [
                        [ id:'test2', run:'1', single_end:false ], // meta map
                        [
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test2_1.fastq.gz', checkifExists: true),
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test2_2.fastq.gz', checkifExists: true)
                        ]
                    ]
                )
                input[1] = params
                """
            }
        }
        then {
            assertAll(
                { assert !workflow.success },
                { assert workflow.stdout.contains("[nf-microbe/toolkit]: --run_bowtie2_host_removal = true but no iGenomes (--genome) or custom fasta (--host_fasta) provided") }
            )
        }
    }
}
