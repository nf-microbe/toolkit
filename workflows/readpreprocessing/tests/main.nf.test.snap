{
    "host_removal: local index input": {
        "content": [
            [
                "versions.yml:md5,127a4928fdfb0700dfdbbee4ff669d91",
                "versions.yml:md5,1ae9481fbb2bf3f462487296b5ab7964",
                "versions.yml:md5,684df08c07c6a52aae4311da0e1ad081",
                "versions.yml:md5,684df08c07c6a52aae4311da0e1ad081"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-03T11:31:37.364833814"
    },
    "host_removal: local fasta input": {
        "content": [
            [
                "versions.yml:md5,127a4928fdfb0700dfdbbee4ff669d91",
                "versions.yml:md5,1ae9481fbb2bf3f462487296b5ab7964",
                "versions.yml:md5,684df08c07c6a52aae4311da0e1ad081",
                "versions.yml:md5,684df08c07c6a52aae4311da0e1ad081",
                "versions.yml:md5,cc09282175c89ea39438e509d5e487ad"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-03T11:32:27.035859924"
    },
    "host_removal: igenomes input": {
        "content": [
            [
                "versions.yml:md5,127a4928fdfb0700dfdbbee4ff669d91",
                "versions.yml:md5,1ae9481fbb2bf3f462487296b5ab7964",
                "versions.yml:md5,684df08c07c6a52aae4311da0e1ad081",
                "versions.yml:md5,684df08c07c6a52aae4311da0e1ad081"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-03T11:30:44.475720685"
    }
}